packages/next/server/render.tsx

```diff
@@ -1741,67 +1745,68 @@ function createFlushEffectStream(
-    let resolved = false
-
-    const closeTag = '</body></html>'
-    const suffixUnclosed = suffix ? suffix.split(closeTag)[0] : null
-
-    const doResolve = (renderStream: ReadableStream<Uint8Array>) => {
-      if (!resolved) {
-        resolved = true
-
-        // React will call our callbacks synchronously, so we need to
-        // defer to a microtask to ensure `stream` is set.
-        resolve(
-          Promise.resolve().then(() => {
-            const transforms: Array<TransformStream<Uint8Array, Uint8Array>> = [
-              createBufferedTransformStream(),
-              flushEffectHandler
-                ? createFlushEffectStream(flushEffectHandler)
-                : null,
-              suffixUnclosed != null
-                ? createPrefixStream(suffixUnclosed)
-                : null,
-              dataStream ? createInlineDataStream(dataStream) : null,
-              suffixUnclosed != null ? createSuffixStream(closeTag) : null,
-            ].filter(Boolean) as any
-
-            return transforms.reduce(
-              (readable, transform) => pipeThrough(readable, transform),
-              renderStream
-            )
-          })
-        )
-      }
-    }
+  const closeTag = '</body></html>'
+  const suffixUnclosed = suffix ? suffix.split(closeTag)[0] : null
+
+  const doResolve = (renderStream: ReadableStream<Uint8Array>) => {
+    // React will call our callbacks synchronously, so we need to
+    // defer to a microtask to ensure `stream` is set.
+    const transforms: Array<TransformStream<Uint8Array, Uint8Array>> = [
+      createBufferedTransformStream(),
+      flushEffectHandler ? createFlushEffectStream(flushEffectHandler) : null,
+      suffixUnclosed != null ? createPrefixStream(suffixUnclosed) : null,
+      dataStream ? createInlineDataStream(dataStream) : null,
+      suffixUnclosed != null ? createSuffixStream(closeTag) : null,
+    ].filter(Boolean) as any
+
+    return transforms.reduce(
+      (readable, transform) => pipeThrough(readable, transform),
+      renderStream
+    )
+  }
 
-    const renderStream: ReadableStream<Uint8Array> = await (
-      ReactDOMServer as any
-    ).renderToReadableStream(element, {
-      onError(err: Error) {
-        if (!resolved) {
-          resolved = true
-          reject(err)
-        }
-      },
+  let completeCallback: (value?: unknown) => void
+  const allComplete = new Promise((resolveError, rejectError) => {
+    completeCallback = execOnce((err: unknown) => {
+      if (err) {
+        rejectError(err)
+      } else {
+        resolveError(null)
+      }
     })
+  })
 
-    doResolve(renderStream)
+  const renderStream: ReadableStream<Uint8Array> = await (
+    ReactDOMServer as any
+  ).renderToReadableStream(element, {
+    onError(err: Error) {
+      completeCallback(err)
+    },
+    onCompleteAll() {
+      completeCallback()
+    },
   })
+
+  if (generateStaticHTML) {
+    await allComplete
+  }
+
+  return doResolve(renderStream)
```

Comment:
Any reason to not just inline `doResolve` here?

Reply:
I thought was to keep grouping that part, but yeah we can definitely do it